addons/binding/org.openhab.binding.rotelra1x/src/main/java/org/openhab/binding/rotelra1x/handler/RotelRa1xHandler.java

```diff
@@ -0,0 +1,337 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.rotelra1x.handler;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.IncreaseDecreaseType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import gnu.io.PortInUseException;
+import gnu.io.RXTXPort;
+import gnu.io.SerialPort;
+import gnu.io.UnsupportedCommOperationException;
+
+/**
+ * The {@link RotelRa1xHandler} handles commands and state changes, and communicates
+ * with the amplifier over the serial port interface..
+ *
+ * @author Marius Bjørnstad - Initial contribution
+ */
+public class RotelRa1xHandler extends BaseThingHandler implements Runnable {
+
+    private final static int BAUD = 115200;
+    private int maximumVolume = 0;
+    private RXTXPort serialPort;
+
+    private boolean exit = false;
+    private volatile boolean power = false;
+
+    private Logger logger = LoggerFactory.getLogger(RotelRa1xHandler.class);
+
+    public RotelRa1xHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        maximumVolume = ((BigDecimal) getThing().getConfiguration().get("maximum-volume")).intValue();
+        exit = false;
+        try {
+            connect();
+        } catch (IOException e) {
+            logger.error("Unable to connect to device", e);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        exit = true;
+        disconnect();
+    }
+
+    private void connect() throws IOException {
+        if (serialPort == null) {
+            String portName = (String) getThing().getConfiguration().get("port");
+            if (portName == null) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Serial port name not configured");
+                throw new IOException("Serial port not configured");
```

Comment:
I'm a bit stuck here. I want the send() and sendForce() (lines 265 & 272) to bail out if connect fails, not continue with trying to write to serialPort (that would cause a NullPointerException). I agree that re-throwing an IOException is not the cleanest way. Returning a boolean (connected or not) would do it, but it seems like an exception is a better tool. Perhaps it shouldn't be send()'s job to call connect() anyway.

I could pull the logging out of connect() -- then I could throw exceptions at will, and log the error and update the status higher up. Still would need to catch & re-throw, to set serialPort to null. But then the logging code has to be repeated in handleCommand() and initialize(), and some detail would be lost (difference between a configuration error / communication error) unless different exceptions were used for each.